activerecord/lib/active_record/associations/association.rb

```diff
@@ -239,7 +239,7 @@ def target_scope
         end
 
         def scope_for_association
-          @enable_scoping ? klass.all : klass.scope_for_association
+          @enable_scoping ? klass.unscoped.all : klass.scope_for_association
```

Comment:
Actually, using `klass.unscoped.all` is not true for this context, this change will break the assertion `assert_equal true, reader.skimmer`.

https://github.com/rails/rails/blob/a81968e4f51ecfa0ecc823ca2330c0cd8e46eca2/activerecord/test/cases/associations/has_many_through_associations_test.rb#L689-L694

Thanks for this PR, but I'll close this in favor of #37523.